target/scripts/helpers/dns.sh

```diff
@@ -7,59 +7,106 @@ function _get_label_count
-  if [[ $(_get_label_count "${HOSTNAME}") -gt 2 ]]
+  # Check whether we're running a bare domain (i.e. if `DMS_FQDN`` is more than 2 labels
+  # long (eg: mail.example.test).
+  if [[ $(_get_label_count "${DMS_FQDN}") -gt 2 ]]
   then
-    if [[ -n ${OVERRIDE_HOSTNAME} ]]
+    # using a subdomain
+    if [[ -n ${DMS_DOMAINNAME+set} ]]
     then
-      # Emulates the intended behaviour of `hostname -d`:
-      # Assign the HOSTNAME value minus everything up to and including the first `.`
-      DOMAINNAME=${HOSTNAME#*.}
+      _log 'trace' "'DMS_DOMAINNAME' supplied"
     else
-      # Operates on the FQDN returned from querying `/etc/hosts` or fallback DNS:
-      #
-      # Note if you want the actual NIS `domainname`, use the `domainname` command,
-      # or `cat /proc/sys/kernel/domainname`.
-      # Our usage of `domainname` is under consideration as legacy, and not advised
-      # going forward. In future our docs should drop any mention of it.
+      _log 'debug' "'DMS_DOMAINNAME' not supplied; the value will be derived"
+      # https://devhints.io/bash#parameter-expansions
+      DMS_DOMAINNAME=${DMS_FQDN#*.}
+    fi
+    if [[ -n ${DMS_HOSTNAME+set} ]]
+    then
+      _log 'trace' "'DMS_HOSTNAME' supplied"
+    else
+      _log 'debug' "'DMS_HOSTNAME' not supplied; the value will be derived"
+      DMS_HOSTNAME=$(cut -d '.' -f 1 <<< "${DMS_FQDN}")
+    fi
+
+    echo "${DMS_HOSTNAME}" >/etc/hostname
+  else
+    # bare domain
+    _log 'debug' 'Detected a bare domain setup'
 
-      #shellcheck disable=SC2034
-      DOMAINNAME=$(hostname -d)
+    # small sanity check
+    if [[ -n ${DMS_HOSTNAME+set} ]]
+    then
+      _log 'warn' "Running a bare domain setup but 'DMS_HOSTNAME' is set - this does not make sense!"
+      _log 'warn' "Emptying 'DMS_HOSTNAME' now"
+    fi
+    DMS_HOSTNAME=
+
+    if [[ -n ${DMS_DOMAINNAME+set} ]]
+    then
+      _log 'trace' "'DMS_DOMAINNAME' supplied"
+    else
+      _log 'debug' "'DMS_DOMAINNAME' not supplied; the value will be derived"
+      DMS_DOMAINNAME=${DMS_FQDN}
     fi
+
+    echo "${DMS_FQDN}" >/etc/hostname
   fi
 
-  # Otherwise we assign the same value (eg: example.com):
-  # Not an else statement in the previous conditional in the event that `hostname -d` fails.
-  DOMAINNAME="${DOMAINNAME:-${HOSTNAME}}"
+  # handle /etc/hosts as well
+  # tools like `openssl` require this to be correctc
+  echo " IP              FQDN (CANONICAL_HOSTNAME)    ALIASES
+# --------------  ---------------------------  -----------------------
+
+127.0.0.1         localhost
+127.0.1.1         ${DMS_FQDN}   ${DMS_HOSTNAME}" >/etc/hosts 2>/dev/null
+
+  if [[ $(tr -d '\n' < /sys/module/ipv6/parameters/disable) -eq 0 ]]
+  then
+    echo "
+
+# The following lines are desirable for IPv6 capable hosts
+::1     ip6-localhost ip6-loopback
+fe00::0 ip6-localnet
+ff00::0 ip6-mcastprefix
+ff02::1 ip6-allnodes
+ff02::2 ip6-allrouters" >>/etc/hosts 2>/dev/null
+  fi
```

Comment:
I'm pretty sure I'm not going to approve of this.

Docker does a fair bit here depending on Docker daemon config, the host itself and how the container is run or any connected networks. We should not mess with it.

Reply:
My current `/etc/hosts` looks like this:

```
127.0.0.1       localhost
::1     localhost ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
172.19.0.2      mail.example.com mail
```

This looks already fine to me. I am not sure, what exactly is achieved by manually creating the hosts file.

Reply:
The rationale here is that one should not need to provide `hostname:` or something similar with Compose or other environments. The ENV should be enough. I can make the changes less intrusive though, and just append something instead of recreating it.